### PR TITLE
Edits to law list, law remove, law edit and law move

### PR DESCRIPTION
**Pull request by Kyrenic**
This pull request should close two issues for milestone 5.1.0, namely #1651 and #1648 .

### CODE REVIEW

**Comment by Kyrenic** [comment]
One thing to note is that we have to clear the law table, I currently see no way to give already existing laws a number index. If anyone has a suggestion, it would be very welcome, but for now I just opted to empty the table instead.

**Review by dmccoystephenson — COMMENTED** [review]
Looks good! Let's just double check that the german character encoding is correct and shows up as intended in-game.

File: src/main/resources/lang/lang_de_DE.properties [hunk 1]
```diff
@@ -157,7 +159,7 @@ CommandFactionCreateNameTooLong=Dein Fraktionsname darf nicht lï¿½nger als {0
 CommandFactionCreateFactionAlreadyExists=Eine Fraktion mit diesem Namen existiert bereits.
 CommandFactionCreateFactionFailedToSave=Ein Fehler ist beim Speichern deiner Fraktion aufgetreten.
 CommandFactionCreateSuccess=Fraktion {0} wurde erstellt.
-CommandFactionLawUsage=Benutze: /Fraktionsgesetz [Hinzufï¿½gen|Entfernen|Liste]
+CommandFactionLawUsage=Benutze: /Fraktionsgesetz [Hinzufï¿½gen|Entfernen|Liste|edit|move]
```
[the remark this answers is not in the document]
Comment (dmccoystephenson) [reply src/main/resources/lang/lang_de_DE.properties:162]: Looks like there are some conflicts to resolve with the german language file. Resolving these might fix this.

File: src/main/kotlin/com/dansplugins/factionsystem/law/JooqMfLawRepository.kt [hunk 2]
```diff
@@ -38,15 +48,56 @@ class JooqMfLawRepository(
     }
 
     override fun delete(id: MfLawId) {
+        delete(getLaw(id)!!)
```
Comment (alyphen) [inline src/main/kotlin/com/dansplugins/factionsystem/law/JooqMfLawRepository.kt:51]: Remember that these operations can occur concurrently due to the async blocks - getLaw may not exist so asserting it does may cause errors.

File: src/main/kotlin/com/dansplugins/factionsystem/law/JooqMfLawRepository.kt [hunk 3]
```diff
@@ -38,15 +48,56 @@ class JooqMfLawRepository(
...
+    }
+
+    override fun delete(law: MfLaw) {
         dsl.deleteFrom(MF_LAW)
-            .where(MF_LAW.ID.eq(id.value))
+            .where(MF_LAW.ID.eq(law.id.value))
+            .execute()
+        dsl.update(MF_LAW)
```
Comment (alyphen) [inline src/main/kotlin/com/dansplugins/factionsystem/law/JooqMfLawRepository.kt:58]: Could this be achieved without updating all the numbers each time? Perhaps use it as a sort index for display and then treat the display number differently?

File: src/main/kotlin/com/dansplugins/factionsystem/law/JooqMfLawRepository.kt [hunk 4]
```diff
@@ -38,15 +48,56 @@ class JooqMfLawRepository(
...
+    override fun edit(law: MfLaw, text: String) {
+        dsl.update(MF_LAW)
+            .set(MF_LAW.TEXT, text)
+            .where(MF_LAW.FACTION_ID.eq(law.factionId.value))
+            .and(MF_LAW.NUMBER.eq(law.number))
+            .execute()
+    }
+
+    override fun move(law: MfLaw, number: Int) {
+        if (law.number!! > number) {
```
Comment (alyphen) [inline src/main/kotlin/com/dansplugins/factionsystem/law/JooqMfLawRepository.kt:73]: Please avoid `!!` if possible

File: src/main/kotlin/com/dansplugins/factionsystem/law/JooqMfLawRepository.kt [hunk 5]
```diff
@@ -38,15 +48,56 @@ class JooqMfLawRepository(
...
-            .where(MF_LAW.ID.eq(id.value))
+            .where(MF_LAW.ID.eq(law.id.value))
+            .execute()
+        dsl.update(MF_LAW)
+            .set(MF_LAW.NUMBER, MF_LAW.NUMBER - 1)
+            .where(MF_LAW.FACTION_ID.eq(law.factionId.value))
+            .and(MF_LAW.NUMBER.gt(law.number))
+    }
+
+    override fun edit(law: MfLaw, text: String) {
```
Comment (alyphen) [inline src/main/kotlin/com/dansplugins/factionsystem/law/JooqMfLawRepository.kt:64]: upsert can be used for this purpose - we don't need a separate query

File: src/main/resources/com/dansplugins/factionsystem/db/migration/V7__Store_law_index.sql [hunk 6]
```diff
@@ -0,0 +1,7 @@
+DELETE FROM `mf_law`;
+
+ALTER TABLE `mf_law` ADD `number` integer NOT NULL;
+
+--CREATE VIEW `lawView` AS SELECT ROW_NUMBER() OVER(PARTITION BY `faction_id`) AS `row_num`, `number` FROM `mf_law` ORDER BY `faction_id`, `row_num`
```
Comment (alyphen) [inline src/main/resources/com/dansplugins/factionsystem/db/migration/V7__Store_law_index.sql:5]: Comments should be removed from this migration before merging.

File: src/main/resources/lang/lang_de_DE.properties (hunk 1 above)
[the remark this answers is not in the document]
Comment (alyphen) [reply src/main/resources/lang/lang_de_DE.properties:162]: I think this is caused by the base branch at the time this branch was created. It's worth doing a rebase.

**Review by alyphen — CHANGES_REQUESTED** [review]

File: src/main/resources/com/dansplugins/factionsystem/db/migration/V7__Store_law_index.sql [hunk 7]
```diff
@@ -0,0 +1,7 @@
+DELETE FROM `mf_law`;
+
+ALTER TABLE `mf_law` ADD `number` integer NOT NULL;
```
Comment (alyphen) [inline src/main/resources/com/dansplugins/factionsystem/db/migration/V7__Store_law_index.sql:3]: Adding a default here or allowing null could prevent us having to empty the table.

**Comment by alyphen** [comment]
> One thing to note is that we have to clear the law table, I currently see no way to give already existing laws a number index. If anyone has a suggestion, it would be very welcome, but for now I just opted to empty the table instead.

I don't think I'm comfortable emptying everybody's law table the moment this update is deployed. If we could assign null or a default value to the number/sort and then assign one during runtime when laws are viewed, that would be preferable I think.

File: src/main/kotlin/com/dansplugins/factionsystem/law/JooqMfLawRepository.kt (hunk 3 above)
Comment (Kyrenic) [reply src/main/kotlin/com/dansplugins/factionsystem/law/JooqMfLawRepository.kt:58]: I don't see how I would do that, but you're welcome to make a commit to my fork if you know how to.

**Comment by Kyrenic** [comment]
The translation errors are still present on my branch, you'll need to walk me through how to rebase it since I don't know how to do that.

I've changed the SQL so it doesn't empty the table anymore, and it now assigns a number index when the laws are viewed if it doesn't have one already.

**Pull request closed without merging**